📈 Elliptic Curve and Modexp Precompiles

ZamDimon opened the pull request:
# What ❔

This pull request adds the evm abstractions for the following precompiles:
- `ecadd`: adding two points on the elliptic curve;
- `ecmul`: multiplying two points on the elliptic curve;
- `ecpairing`: verification of pairing operation;
- `modexp`: modular exponentiation.

## Why ❔

They haven't been implemented yet, thus the PR.

## Checklist

- [x] PR title corresponds to the PR body (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `zk fmt` and `zk lint`.

ZamDimon committed:
:tada: add `ecadd` precompile implementation

ZamDimon committed:
:sparkles: add `ecmul` precompile

ZamDimon committed:
:sparkles: add `ecpairing` precompile

ZamDimon committed:
:test_tube: add tests for `ecadd` and `ecmul`

ZamDimon committed:
:adhesive_bandage: :art: do `cargo fmt` and fix comments in tests

ZamDimon committed:
:sparkles: :test_tube: finally figure out `ecpairing` EIP spec

ZamDimon committed:
:sparkles: add `modexp` precompile

ZamDimon committed:
:sparkles: make more rounds in `ecpairing`

ZamDimon committed:
:memo: fix comments in `ecpairing`

ZamDimon committed:
:adhesive_bandage: fix `write\_history` in `ecadd` precompile abstraction

ZamDimon committed:
:bug: fix `ecmul` write history

ZamDimon committed:
:bug: fix typo in `modexp`

ZamDimon committed:
:bug: fix typo in `ecpairing`

jules commented on src/vm.rs:
```diff
@@ -214,3 +224,12 @@ pub trait Precompile: std::fmt::Debug {
         Option<(Vec<MemoryQuery>, Vec<MemoryQuery>, Vec<Self::CycleWitness>)>,
     );
 }
+
+pub enum SpongeExecutionMarker {
+    MemoryQuery,
+    DecommittmentQuery,
+    StorageLogReadOnly,
+    StorageLogWrite,
+    CallstackPush,
+    CallstackPop,
+}
```
all looks good, but did this maybe pop up since you based off a different branch?

nikitamasych replied:
Yep, rebased to `dl-precompiles`

vladbochok commented on src/precompiles/ecmul.rs:
```diff
@@ -0,0 +1,357 @@
+    // and validating that the conversion is successful
+    let x1_field = Fq::from_str(x1.to_string().as_str()).ok_or(Error::msg("invalid x1"))?;
+    let y1_field = Fq::from_str(y1.to_string().as_str()).ok_or(Error::msg("invalid y1"))?;
+    let s_field = Fr::from_str(s.to_string().as_str()).ok_or(Error::msg("invalid s"))?;
```
Why is that? `s` can be an arbitrary uint256 number

nikitamasych replied:
True, missed that, we will update

ZamDimon replied:
Yeah, indeed, missed that. Now, `s` can be arbitrary. Also, I added three different tests where the scalar is greater than the group order to make sure we have not missed anything

nikitamasych committed:
:art: cargo fmt + Cargo.toml update opcodes version

ZamDimon committed:
:bug: make scalar input to `ecmul` possible outside the field

ZamDimon committed:
:bug: fix point at infinity bad representation

ZamDimon commented:
Fixed the issue where the point at infinity was encoded as a tuple `(0,1)` due to the internal implementation in https://github.com/matter-labs/pairing.

According to the precompile spec, the point at infinity is now encoded as `(0,0)` (see [here](https://www.evm.codes/precompiled#0x07). I have added corresponding tests to assert that this now holds.

ZamDimon committed:
:test_tube: add tests to cover cases with the points at infinity

nikitamasych committed:
:art: cargo fmt

nikitamasych committed:
:pushpin: pin to git version instead of the local path in Cargo.toml
